### PR TITLE
Add Legacy CUDA ROCm Jobs w/ Docs

### PR DESCRIPTION
**Pull request by walshmm**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes
Add ROCm jobs for legacy CUDA for issue #3914

## What type(s) of changes does this code introduce?
Adds self hosted jobs for nitrogen ROCm Legacy Cuda Jobs

During testing, there were some test failures on nitrogen:

```
ROCm-Clang13-NoMPI-Legacy-CUDA2HIP-Real
Failures:
None

ROCm-Clang13-NoMPI-Legacy-CUDA2HIP-Real-Mixed
Failures:
397 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-flux (Failed)

ROCm-Clang13-NoMPI-Legacy-CUDA2HIP-Complex
Failures:
None

ROCm-Clang13-NoMPI-Legacy-CUDA2HIP-Complex-Mixed
Failures:
397 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-kinetic (Failed)
401 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-potential (Failed)
403 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-localecp (Failed)
```
Should this be moved to an issue?

- Other (please describe): Github Actions Job

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

File: docs/github_actions.rst
Comment (ye-luo) [inline docs/github_actions.rst:781]: This line is still needed.

**Comment by ye-luo** [comment]
May be this PR can be used to demonstrate the rebase feature?

**Comment by walshmm** [comment]
@ye-luo 
> May be this PR can be used to demonstrate the rebase feature?

Sure!

**Comment by prckent** [comment]
!-> Feel free to automatically rebase this PR. <-!

**Comment by walshmm** [comment]
> !-> Feel free to automatically rebase this PR. <-!
@prckent 
It only checks the pr description for this line.  I have added it already which triggered [this](https://github.com/QMCPACK/qmcpack/actions/runs/2258757207).  
Seems like it didnt do any rebasing yet as expected :+1:

**Review by ye-luo — DISMISSED** [review]
LGTM

**Comment by ye-luo** [comment]
Test this please

**Comment by prckent** [comment]
(Now this is up to date,  rebaser won't do anything. Is there another PR ready to approve+merge?)

**Comment by walshmm** [comment]
> (Now this is up to date, rebaser won't do anything. Is there another PR ready to approve+merge?)
@prckent 
I can force push my local branch to undo the merge if you would like.

**Comment by PDoakORNL** [comment]
If this one is merged it could be tried with my #3996

**Comment by ye-luo** [comment]
I merged develop into this PR branch, and thus the base should kick in and remove the merge commit 0bd5182 I assume

**Comment by walshmm** [comment]
> @walshmm `.github/workflows/ci-github-action-auto-rebase.yaml / rebase` CI failed. What does this workflow do? If it fails, will the merge get blocked?

@ye-luo 

This is the workflow that handles the rebasing.  As for why it failed it seemed to not be able to find one of the scripts(odd because it worked fine in my fork and Mark's fork).  

Currently it will not block the merge on failure,  I think that would require making these jobs required.  Though the only times it is expected to fail when functioning correctly is when a branch is not rebaseable (I think the only situation that is in is if there is a merge conflict)

I'll have to investigate why it failed, if its a network hiccup then we may need some retry functionality.

**Comment by ye-luo** [comment]
@walshmm "This is the workflow that handles the rebasing." I'm still very confused.
It seems being triggered when auto-merge is enabled. What does this pipeline do? Test if a rebase can be executed? Does the rebase actually still happen at the final merge? What are the steps in this workflow? In case the rebase CI succeed while the merge has not yet happened due to waiting for other CI runs. What is the outcome? Could you break up the steps to help me understanding. Once I understand the steps, I maybe can self answer my questions.

"Enable auto-merge" also has an option to "rebase and merge"
if PR author adds rebase request, and PR approval select enable auto-merge, what happens?
if PR author adds rebase request, and PR approval select enable auto-merge with "squash and merge", what happens?
if PR author adds rebase request, and PR approval select enable auto-merge with "rebase and merge", what happens?

**Comment by walshmm** [comment]
@ye-luo 

> @walshmm "This is the workflow that handles the rebasing." I'm still very confused. It seems being triggered when auto-merge is enabled.
> What does this pipeline do? 

The pipeline itself runs if automerge is enabled or if there is a push to either develop/main.  In the second case it checks for open PR's that are set to automerge, and pushes a rebase.

> Test if a rebase can be executed?

Most of the rebase work is handled in [this script](https://github.com/QMCPACK/qmcpack/blob/develop/external_codes/github_actions/auto_rebase.sh)
It tests if it can be rebased with a few retries in case of networking issues, and if i can it will rebase.  Otherwise it will not alter anything.

> Does the rebase actually still happen at the final merge?

What do you mean by this?  Like the one that github does if you when you hit "Merge Pull Request" or are you asking about the timing of the rebase job?

>What are the steps in this workflow? In case the rebase CI succeed while the merge has not yet happened due to waiting for other CI runs. What is the outcome? Could you break up the steps to help me understanding. Once I understand the steps, I maybe can self answer my questions.
> 

If a rebase is necessary and succeeds while other tests are still running, this will cause a push to the incoming branch.  A push to the incoming branch should retrigger tests.(much like what currently happens manually)


There are 2 main cases for when a rebase job is triggered:

* Automerge Enabled
  * To enable an automerge a pr must be approved.  This means it will only rebase when necessary, and long lived PR's wont use resources until ready.
  * On automerge enable with the permission message included in the PR description, the workflow will do the following:
    * Checkout target branch
    * Execute `auto-rebase.sh` (linked above):
      * Initialize with github actions env variables such as `$GITHUB_TOKEN`
      * Check rebase status
      * If rebasable and not up to date, perform a rebase
      * If successful, push rebase to target branch
      * If a need for rebase cannot be determined or is needed and fails, exit code 1 and fails the job
* Push to a Base Branch
  * Checkout Base branch so that we may access the rebase scripts
  * Execute [trigger-rebase-on-push.sh](https://github.com/QMCPACK/qmcpack/blob/develop/external_codes/github_actions/trigger-rebase-on-push.sh)
    * Collect list of open PR's
    * For each PR that has automerge active and includes the permission message, trigger the `auto_rebase.sh` script as above.



> "Enable auto-merge" also has an option to "rebase and merge" if PR author adds rebase request, and PR approval select enable auto-merge, what happens? if PR author adds rebase request, and PR approval select enable auto-merge with "squash and merge", what happens? if PR author adds rebase request, and PR approval select enable auto-merge with "rebase and merge", what happens?

The rebase job/script is independent of this github feature.   Regardless of choice the job will just rebase the incoming branch, and then the method you choose to actually merge with will take place when all required jobs finish.

I think the main point of this job was so that the code was rebased for the required tests so that they accurately test the next proposed state of develop.
It may make sense to have the required test jobs depend on a successful run of rebase to start(either success from it already being up to date or a successful rebase).  This will ensure it is rebased before a merge and the tests actually benefit from it on first run.  The risk to that is stalling pr's if there were any issues with the rebase script, but I supposed a "force success" flag could circumvent that.

**Comment by prckent** [comment]
(Following from and in agreement with @walshmm 's comments, an example use case is that we could eventually get 4 approved PRs merged one evening without requiring human intervention.  This will also make it easier to handle many smaller PRs. However, further maturation is needed once the script is matured in its current state - the triggering of the internal CI will need to be worked out before human intervention is not required.)

**Comment by ye-luo** [comment]
There are 2 main cases for when a rebase job is triggered:
- Automerge Enabled
- Push to a Base Branch

The rebase job/script is independent of github auto-merge feature. 

These are the info I was looking for. It seems clear to me know.
For the end goal @prckent mentioned. I think more steps will be needed.
I think the in-house CI right now requires a message "Test this please". Once a rebase job gets executed successfully, the lack of  "Test this please" stalls the PR.

**Comment by ye-luo** [comment]
@walshmm auto_rebase.sh vs auto-rebase.sh

**Comment by walshmm** [comment]
@ye-luo 
> @walshmm auto_rebase.sh vs auto-rebase.sh

Ah yes thank you, must have made that typo creating the clean branch.  
#3998 pr to fix the name.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Rebase worked.

**Comment by walshmm** [comment]
hmm it seems the rebase staled the review but auto-merge is still enabled.  If we want everything to be automatic we will probably need to have the bot reapprove after a rebase if thats possible.

**Comment by ye-luo** [comment]
github cloud CI doesn't go after rebasing.

**Comment by ye-luo** [comment]
Here is my suggestion on this PR. We need the legacy CUDA2HIP CI. Thus let us disable the rebase and merge this PR. Further testing of the rebase feature can be done with a new fake PR.

**Comment by walshmm** [comment]
> Here is my suggestion on this PR. We need the legacy CUDA2HIP CI. Thus let us disable the rebase and merge this PR. Further testing of the rebase feature can be done with a new fake PR.

Alrighty, disabled for this pr

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**
